src/Lecture3.hs

```diff
@@ -181,7 +230,13 @@ together only different elements.
 Product {getProduct = 6}
 
 -}
-appendDiff3 = error "TODO"
+
+appendDiff3 :: (Semigroup a, Eq a) => a -> a -> a -> a
+appendDiff3 x y z
+  | x == y && y == z = x
+  | x == y && y /= z = y <> z
+  | x /= y && (y == z || x == z) = x <> y
+  | x /= y && x /= z = x <> y <> z
```

Comment:
You have non-exhaustive pattern matching here because GHC can't figure out if your equations cover all possible situations. The last case in guards should always be `otherwise` to make GHC happy 🙂 

You can implement this function in a more elegant way if you have a look at functions `nub` and `foldl1'`

Reply:
I know, but nub is O^2 in the time complexity right, which means this is the more efficient solution ?

Reply:
Yes, your solution is more efficient at the moment 👍🏻 
`O(n^2)` is not a problem for the list of 3 elements. In real life, you can use something like `ordNub` that has `O(n log n)` complexity